### PR TITLE
[Dotenv] Reimplementing symfony/flex' dump-env as a Symfony command

### LINKED ISSUE #40381: Provide "dump-env" as regular Symfony command
Symfony Flex provides a `dump-env` command which must be invoked using Composer. At [its introduction](https://symfony.com/blog/new-in-symfony-flex-1-2) it was suggested to use this in deployment workflows to generate optimized PHP versions of `.env`.

However, this assumes that Composer exists and is executed on remote servers. At least for us this is impossible since our deployment workflows dictate that releases are prepared locally (installing Composer dependencies, etc.) before being uploaded to the remote server. Thus we don't need/have Composer on the remote server.

Since I assume that we are not alone I would suggest to provide `dump-env` also as regular Symfony command which can be invoked using `bin/console`.

### PR DESCRIPTION
**Pull request by abdielcs**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #40381
| License       | MIT
| Doc PR        | -

[FrameworkBundle] Reimplementing the symfony/flex env-dump as a symfony command. Most of code copied from https://github.com/symfony/flex/blob/main/src/Command/DumpEnvCommand.php

The command is not registered by default. In order to enable it, one must add it to their `services.yaml` file:
```yaml
services:
    Symfony\Component\Dotenv\Command\DotenvDumpCommand:
        - '%kernel.project_dir%/.env'
        - '%kernel.environment%'
```

On PHP >= 8, the two arguments can be removed when autoconfiguration is enabled (which is the default):
```yaml
services:
    Symfony\Component\Dotenv\Command\DotenvDumpCommand: ~
```

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Command/DumpEnvCommand.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+use Symfony\Component\Dotenv\Dotenv;
+
+/**
+ * A console command to compiles the contents of the .env* files into a PHP-optimized file called .env.local.php.
+ *
+ * @internal
+ */
+class DumpEnvCommand extends Command
+{
+    protected static $defaultName = 'env:dump';
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/DumpEnvCommand.php]: `dotenv:dump` IMHO

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 2]
```diff
@@ -3,7 +3,7 @@ CHANGELOG
 
 5.4
 ---
-
+ * Added `env:dump` command to compiles the contents of the .env* files into a PHP-optimized file called .env.local.php. Copied from symfony/flex.
```
Comment (t-richard) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]:
```suggestion
 * Add `env:dump` command to compile the contents of the .env* files into a PHP-optimized file called `.env.local.php`. Copied from symfony/flex.
```

File: src/Symfony/Bundle/FrameworkBundle/Command/DumpEnvCommand.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Dotenv\Dotenv;
+
+/**
+ * A console command to compiles the contents of the .env* files into a PHP-optimized file called .env.local.php.
```
Comment (t-richard) [inline src/Symfony/Bundle/FrameworkBundle/Command/DumpEnvCommand.php]:
```suggestion
 * A console command to compile the contents of the .env* files into a PHP-optimized file called .env.local.php.
```

**Review by t-richard — CHANGES_REQUESTED** [review]
Its also missing tests. See https://github.com/symfony/flex/blob/main/tests/Command/DumpEnvCommandTest.php

Otherwise I'm :+1:, it also proved to be sub-optimal for me to have it as a composer plugin command rather than a symfony command.

It would be great if you were also able to submit a PR to https://github.com/symfony/flex for the deprecation of the the command.

File: src/Symfony/Bundle/FrameworkBundle/Command/DumpEnvCommand.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * @internal
+ */
+class DumpEnvCommand extends Command
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Command/DumpEnvCommand.php]:
```suggestion
final class DumpEnvCommand extends Command
```

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: s/Added/Add also

**Comment by abdielcs** [comment]
Help with this issues please? `unset($_ENV['SYMFONY_DOTENV_VARS'])` came from original code in https://github.com/symfony/flex/blob/main/src/Command/DumpEnvCommand.php and need to be done to pass tests, and Psalm insists that is not set. Not sure what need to be done to rebase upstream. Seems also a test is failing from HttpKernel.

**Comment by ro0NL** [comment]
See https://github.com/vimeo/psalm/issues/6337

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 5]
```diff
@@ -3,7 +3,7 @@ CHANGELOG
 
 5.4
 ---
-
+ * Add `dotenv:dump` command to compile the contents of the .env* files into a PHP-optimized file called `.env.local.php`. Copied from symfony/flex.
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: s/.env*/.env (remove the `*`)

File: src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Dotenv\Dotenv;
+
+/**
+ * A console command to compile the contents of the .env* files into a PHP-optimized file called .env.local.php.
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php]: s/.env*/.env (remove the `*`)

File: src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * @internal
+ */
+final class DotenvDumpCommand extends Command
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php]: I just realized that this should be in the Dotenv component, isn't it?
Can you please move it there?

**Comment by nicolas-grekas** [comment]
Please rebase and squash also (PR should not contain any merge commits)

File: src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php (hunk 7 above)
Comment (abdielcs) [reply src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php]: But that would mean making the Dotenv component dependent on symfony/console (not only on require-dev it seems to me). Would that be ok? Also, will everyone who uses Dotenv  have the need to export these variables in the same place?

**Comment by nicolas-grekas** [comment]
We might need to implement https://github.com/symfony/flex/pull/810 in this also.
@abdielcs are you still available to finish this PR?

**Comment by abdielcs** [comment]
> We might need to implement [symfony/flex#810](https://github.com/symfony/flex/pull/810) in this also.
> @abdielcs are you still available to finish this PR?

@nicolas-grekas  Al the work is done, but I can't managed to pass the rebase. Really would appreciate help here. Sorry for been so newbie doing PR.

File: src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Command/DotenvDumpCommand.php]: Optional feature don't need to be in "require". Putting "symfony/console" in "require-dev" is how we should do this.

**Comment by nicolas-grekas** [comment]
@abdielcs I rebased the PR and pushed it on your fork. Please fetch and sync your local copy before moving the command to the component.

**Review by nicolas-grekas — APPROVED** [review]
I rebased the PR and added a second commit to completly decouple the command from FrameworkBundle.

@abdielcs I saw that you somehow messed up with the git history in previous iterations. As a general rule of thumb, do never ever use `git merge` when working on PRs for open-source projects. Use `git rebase` instead.

**Review by derrabus — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @abdielcs.

**Comment by mbrodala** [comment]
Awesome thanks a lot!

**Pull request merged**